### ISSUE TITLE
Infinite recursion inside the assertion rewrite import hooks

### ISSUE THREAD

**Issue by ionelmc** [pytest-dev/pytest#8461]
I suspect it's some bad interaction with eventlet (with it's own import hooks I guess?)

```
GLOB sdist-make: /home/ionel/open-source/python-manhole/setup.py
py27-normal-eventlet-nocov inst-nodeps: /home/ionel/open-source/python-manhole/.tox/.tmp/package/4/manhole-1.6.0.zip
py27-normal-eventlet-nocov installed: DEPRECATION: Python 2.7 reached the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 is no longer maintained. pip 21.0 will drop support for Python 2.7 in January 2021. More details about Python 2 support in pip can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support pip 21.0 will remove support for this functionality.,atomicwrites==1.4.0,attrs==20.3.0,backports.functools-lru-cache==1.6.1,certifi==2020.12.5,chardet==4.0.0,configparser==4.0.2,contextlib2==0.6.0.post1,dnspython==1.16.0,enum34==1.1.10,eventlet==0.24.1,funcsigs==1.0.2,greenlet==1.0.0,idna==2.10,importlib-metadata==2.1.1,manhole @ file:///home/ionel/open-source/python-manhole/.tox/.tmp/package/4/manhole-1.6.0.zip,monotonic==1.5,more-itertools==5.0.0,packaging==20.9,pathlib2==2.3.5,pluggy==0.13.1,process-tests==2.1.1,py==1.10.0,pyparsing==2.4.7,pytest==4.6.11,pytest-travis-fold==1.3.0,requests==2.25.1,scandir==1.10.0,six==1.15.0,subprocess32==3.5.4,urllib3==1.26.4,uWSGI==2.0.17.1,wcwidth==0.2.5,zipp==1.2.0
py27-normal-eventlet-nocov run-test-pre: PYTHONHASHSEED='488280098'
py27-normal-eventlet-nocov run-test: commands[0] | pytest -vv --ignore=src
=============================================================================================== test session starts ===============================================================================================
platform linux2 -- Python 2.7.18, pytest-4.6.11, py-1.10.0, pluggy-0.13.1 -- /home/ionel/open-source/python-manhole/.tox/py27-normal-eventlet-nocov/bin/python
cachedir: .tox/py27-normal-eventlet-nocov/.pytest_cache
rootdir: /home/ionel/open-source/python-manhole, inifile: setup.cfg, testpaths: tests
plugins: travis-fold-1.3.0
collected 45 items / 1 errors / 44 selected

===================================================================================================== ERRORS ======================================================================================================
_________________________________________________________________________________________ ERROR collecting tests/wsgi.py __________________________________________________________________________________________
tests/wsgi.py:6: in <module>
    import manhole
.tox/py27-normal-eventlet-nocov/lib/python2.7/site-packages/manhole/__init__.py:52: in <module>
    _ORIGINAL_FDOPEN = _get_original('os', 'fdopen')
.tox/py27-normal-eventlet-nocov/lib/python2.7/site-packages/manhole/__init__.py:42: in _get_original
    return getattr(_original(mod), name)
.tox/py27-normal-eventlet-nocov/lib/python2.7/site-packages/eventlet/patcher.py:195: in original
    real_mod = __import__(modname, {}, {}, modname.split('.')[:-1])
.tox/py27-normal-eventlet-nocov/lib/python2.7/site-packages/_pytest/assertion/rewrite.py:91: in find_module
    if self._early_rewrite_bailout(name, state):
.tox/py27-normal-eventlet-nocov/lib/python2.7/site-packages/_pytest/assertion/rewrite.py:213: in _early_rewrite_bailout
    if fnmatch_ex(pat, path):
.tox/py27-normal-eventlet-nocov/lib/python2.7/site-packages/_pytest/pathlib.py:375: in fnmatch_ex
    return fnmatch.fnmatch(name, pattern)
/usr/lib64/python2.7/fnmatch.py:40: in fnmatch
    import os
.tox/py27-normal-eventlet-nocov/lib/python2.7/site-packages/_pytest/assertion/rewrite.py:91: in find_module
    if self._early_rewrite_bailout(name, state):
E   RuntimeError: maximum recursion depth exceeded while calling a Python object
!!! Recursion detected (same locals & position)
================================================================================================ warnings summary =================================================================================================
.tox/py27-normal-eventlet-nocov/lib/python2.7/site-packages/dns/hash.py:25
  /home/ionel/open-source/python-manhole/.tox/py27-normal-eventlet-nocov/lib/python2.7/site-packages/dns/hash.py:25: DeprecationWarning: dns.hash module will be removed in future versions. Please use hashlib instead.
    DeprecationWarning)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
============================================================================================= short test summary info =============================================================================================
ERROR tests/wsgi.py - RuntimeError: maximum recursion depth exceeded while calling a Python object
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
======================================================================================= 1 warnings, 1 error in 0.21 seconds =======================================================================================
In atexit handler.
ERROR: InvocationError for command /home/ionel/open-source/python-manhole/.tox/py27-normal-eventlet-nocov/bin/pytest -vv --ignore=src (exited with code 2)
_____________________________________________________________________________________________________ summary _____________________________________________________________________________________________________
ERROR:   py27-normal-eventlet-nocov: commands failed
```

I'll try to get a simpler reproducer if this doesn't clue into what's wrong.

**Comment by RonnyPfannschmidt** [comment]
i would recommend to just add a self-loop detection and raise a error if there is more than 2 self-loops

**Comment by asottile** [comment]
this is only possible in python 2.x where the `os` import is deferred in `fnmatch.fnmatch` -- as we no longer support python 2.x I don't think there's anything actionable here

this was fixed in 3.2a1 via cc14320159d in cpython
